src/cpu/aarch64/acl_inner_product.cpp

```diff
@@ -48,24 +57,234 @@ status_t acl_inner_product_fwd_t::execute_forward(const exec_ctx_t &ctx) const {
+            attr()->fpmath_.mode_, fpmath_mode::bf16, fpmath_mode::any);
+
+    CHECK(post_ops.init(
+            engine, attr_.post_ops_, dst_md_, aip_.fc_info.activation_info));
+    aip_.use_dst_acc_for_sum = post_ops.has_sum();
+
+    // WeightFormat::ANY tells ACL we can handle any format
+    aip_.weights_info = arm_compute::WeightsInfo(
+            false, 1, 1, ic_total, false, arm_compute::WeightFormat::ANY);
+
+    // Get the format that the ACL kernel will expect the weights to be
+    // in (if a kernel exists) Note that these are referred to as fixed
+    // format kernels, because they require one specific weights format
+    arm_compute::WeightFormat expected_weight_format;
+    ACL_CHECK_VALID(
+            arm_compute::experimental::op::CpuFullyConnected::has_opt_impl(
+                    expected_weight_format, &aip_.src_tensor_info,
+                    &aip_.wei_tensor_info,
+                    aip_.with_bias ? &aip_.bia_tensor_info : nullptr,
+                    &aip_.dst_tensor_info, aip_.fc_info, aip_.weights_info));
+
+    // Set weights info to the one returned by has_opt_impl
+    aip_.weights_info.set_weight_format(expected_weight_format);
+
+    // has_opt_impl may return a non fast math kernel, even if requested
+    aip_.fc_info.enable_fast_math
+            = arm_compute::is_fixed_format_fast_math(expected_weight_format);
+
+    // Inner product is the same as the matmul n x (chw) * (ihw) x o
+    // (note that the src c and weights i both correspond to the input
+    // channel). ACL FullyConnectedLayer assumes the chw dimensions of
+    // src and ihw dimensions of weights are collapsed, so we need to
+    // make sure that they have the same layout. Given that weights are
+    // more often fixed, (so reorders can be hoisted) it makes sense to
+    // reorder the weights to fit the src.
+
+    // For 4D tensors we need to:
+    // - reorder the ihw of the weights to match the src chw
+    // - collapse ihw
+    // - pad the collapsed ihw
+    // But there is not yet a way to express this collapse+pad as a
+    // reorder. So we try to reorder the weights to match the src,
+    // implicitly collapse ihw in our definition of the weights
+    // TensorInfo and hope that the inner_dim has zero padding
+    // (weights_md_.dims[inner_dim] % block_by == 0). If it does, we
+    // fall back to a kernel without blocking (currently this is
+    // equivalent to non-fastmath).
+
+    // 2D just works because we just pad the only dimension.
+
+    // o_dim is always the first logical dimension (oihw, ohwi, oi)
+    dim_t o_dim = 0;
+    dim_t inner_dim;
+    // Rest of logical dimensions in order of innermost to outermost
+    std::vector<dim_t> remaining_dims = {};
+
+    if (src_tag == nchw) {
+        inner_dim = 3; // w
+        remaining_dims = {2, 1}; // h, i
+    } else if (src_tag == nhwc) {
+        inner_dim = 1; // i
+        remaining_dims = {3, 2}; // w, h
+    } else { // Only remaining case is 2D (nc)
```

Comment:
Minor: I'd rather put `nc` under `else if` and assert under `else` for easier debugging.